src/github.com/matrix-org/dendrite/syncserver/storage/output_room_events_table.go

```diff
@@ -47,8 +47,10 @@ const selectRecentEventsSQL = "" +
 const selectMaxIDSQL = "" +
 	"SELECT MAX(id) FROM output_room_events"
 
+// In order for us to apply the state updates correctly, rows need to be ordered in the order they were received (id).
 const selectStateInRangeSQL = "" +
-	"SELECT event_json, add_state_ids, remove_state_ids FROM output_room_events WHERE id > $1 AND id < $2 AND add_state_ids IS NOT NULL OR remove_state_ids IS NOT NULL"
+	"SELECT event_json, add_state_ids, remove_state_ids FROM output_room_events WHERE id > $1 AND id < $2 AND " +
+	"add_state_ids IS NOT NULL OR remove_state_ids IS NOT NULL ORDER BY id ASC"
```

Comment:
Style nit. I'd put the space " " character on the next line because it's easier to read.

so:

```go
     "SELECT ..." +
     " add_state_ids"
```

rather than:

```go
     "SELECT ... " +
     "add_state_ids"
```